temporal-testing/src/main/java/io/temporal/internal/testservice/TestWorkflowMutableStateImpl.java

```diff
@@ -1149,9 +1151,21 @@ private void processFailWorkflowExecution(
           failureType = Optional.of(failureInfo.getType());
           backoffInterval = rs.getBackoffIntervalInSeconds(failureType, store.currentTime());
         }
-      } else {
+      } else if (d.getFailure().hasTerminatedFailureInfo()
```

Comment:
This needs some serious comment explaining what is going on here and why those checks are there.